python/examples/example-ops/softmax-xent/softmax_xent.py

```diff
@@ -0,0 +1,124 @@
+import os
+import torch
+import triton
+
+
+class _softmax_xent_loss(torch.autograd.Function):
+    fwd_src = open(os.path.join(os.path.dirname(__file__), "softmax_xent_fwd.c")).read()
+    bwd_src = open(os.path.join(os.path.dirname(__file__), "softmax_xent_bwd.c")).read()
+
+    # Need TILE = n_vocab for this approach to work:
+    input_config_to_kernel_fwd = {}
+    input_config_to_kernel_bwd = {}
+
+    @classmethod
+    def forward(cls, ctx, logits, indices):
+        n_vocab = logits.shape[-1]
+        assert indices.dtype == torch.int64
+        assert n_vocab % 16 == 0, "Number of logit options must be divisible by 16."
+
+        if not (logits.dtype, n_vocab) in cls.input_config_to_kernel_fwd:
+            cls.input_config_to_kernel_fwd[(logits.dtype, n_vocab)] = triton.kernel(
+                cls.fwd_src,
+                defines={"TILE": n_vocab, "TYPE": logits.dtype},
+                num_warps=[2, 4, 8, 16],
+            )
+        kernel_fwd = cls.input_config_to_kernel_fwd[(logits.dtype, n_vocab)]
+
+        N = logits.numel()
+        result = torch.empty_like(indices, dtype=logits.dtype).cuda()
+        grid = lambda opt: (triton.cdiv(N, opt.d("TILE")),)
+        kernel_fwd(logits, indices, result, grid=grid)
+        # logits -> neg_logprobs via an in place modification by kernel_fwd
+        ctx.save_for_backward(logits, indices)
+
+        return result
+
+    @classmethod
+    def backward(cls, ctx, dneg_logprobs):
+        """We know d(-log(p[i])/dlogit[k] = -id_mat[i,k] + p[k]
+        so we initialize the gradient as neg_logprobs, so we can just exponentiate
+        to get p[k], which is most of what we need...  neg_logprobs will be
+        modified in place to become the gradient we want
+        """
+        neg_logprobs, indices = ctx.saved_tensors
+        assert indices.dtype == torch.int64
+        n_vocab = neg_logprobs.shape[-1]
+        N = neg_logprobs.numel()
+        useful_int = torch.zeros_like(indices).int()
+
+        if not (neg_logprobs.dtype, n_vocab) in cls.input_config_to_kernel_bwd:
+            cls.input_config_to_kernel_bwd[
+                (neg_logprobs.dtype, n_vocab)
+            ] = triton.kernel(
+                cls.bwd_src,
+                defines={"TILE": n_vocab, "TYPE": neg_logprobs.dtype},
+                num_warps=[16],
+            )
```

Comment:
FYI @ptillet this is the place where we had to set num_warps=[16], or else we get incorrect results when running `test_grad` on A100s.

Reply:
Do you still get errors when you set `num_warps = [4]` ? I tried it on an RTX3070 and couldn't see a failure. What are the values of the incorrect tensor in this case?

Reply:
You're right, the code seems to work for any fixed value = 4, 8, 16.